manual/CHAPTER_TextRtlil.tex

```diff
@@ -0,0 +1,270 @@
+    \item \texttt{z}: A high-impedance value (or don't care in case patterns)
+    \item \texttt{m}: A marked bit (internal use only)
+    \item \texttt{-}: A don't care value
+\end{itemize}
+
+An \textit{integer} is simply a signed integer value in decimal format.
+
+\begin{indentgrammar}{<binary-digit>}
+<value> ::= <decimal-digit>$+$ \texttt{\textbf{'}} <binary-digit>$*$
+
+<binary-digit> ::= "0" | "1" | "x" | "z" | "m" | "-"
+
+<integer> ::= "-"$?$ <decimal-digit>$+$
+\end{indentgrammar}
+
+\subsection{Strings}
+
+A string is a series of characters delimited by double-quote characters. Within a string, certain escapes can be used:
+
+\begin{itemize}
+    \item \texttt{\textbackslash n}: A newline
+    \item \texttt{\textbackslash t}: A tab
+    \item \texttt{\textbackslash \textit{ooo}}: A character specified as a one, two, or three digit octal value
+\end{itemize}
+
+All other characters may be escaped by a backslash, and become the following character. Thus:
+
+\begin{itemize}
+    \item \texttt{\textbackslash \textbackslash}: A backslash
+    \item \texttt{\textbackslash \"}: A double-quote
+    \item \texttt{\textbackslash r}: An 'r' character
+\end{itemize}
+
+\subsection{Comments}
+
+A comment starts with a \texttt{\textbf{\#}} character and proceeds to the end of the line. All comments are ignored.
+
+\section{File}
+
+A file consists of an optional autoindex statement followed by zero or more modules.
+
+\begin{indentgrammar}{<design>}
+<file> ::= <autoidx-stmt>$?$ <module>*
+\end{indentgrammar}
+
+\subsection{Autoindex statements}
+
+The autoindex statement sets the global autoindex value used by Yosys when it needs to generate a unique name, e.g. \texttt{\textdollar{}flatten\textdollar{}N}. The N part is filled with the value of the global autoindex value, which is subsequently incremented. This global has to be dumped into RTLIL, otherwise e.g. dumping and running a pass would have different properties than just running a pass on a warm design.
+
+\begin{indentgrammar}{<autoidx-stmt>}
+<autoidx-stmt> ::= "autoidx" <integer> <eol>
+\end{indentgrammar}
+
+\subsection{Modules}
+
+Declares a module, with zero or more attributes, consisting of zero or more wires, memories, cells, processes, and connections.
+
+\begin{indentgrammar}{<module-body-stmt>}
+<module> ::= <attr-stmt>$*$ <module-stmt> <module-body> <module-end-stmt>
+
+<module-stmt> ::= "module" <id> <eol>
+
+<module-body> ::= 
+(<param-stmt>
+  \alt <wire>
+  \alt <memory>
+  \alt <cell>
+  \alt <process>
+  \alt <connection>)$*$
+
+<param-stmt> ::= "parameter" <id> <constant>$?$ <eol>
+
+<constant> ::= <value> | <integer> | <string>
+
+<module-end-stmt> ::= "end" <eol>
+\end{indentgrammar}
+
+\subsection{Attribute statements}
+
+Declares an attribute with the given identifier and value.
+
+\textbf{Warning:} There is currently a bug where integer constants are silently truncated to 32 bits and treated as unsigned.
```

Comment:
This isn't a bug, it's intended behavior (I asked). It's just obnoxious and IMO must be changed.

Reply:
Oof. Just to be clear: you can still have negative constants, and they will remain negative, unless the constant is greater than 32 bits?

Reply:
No, it's much weirder than that. For bit vectors (like `64'00000000...0000`) there is no problem, they get treated as bit vectors. (Incidentally you can't specify sign.) The problems only arise for bare integers (like `12345`) that are a part of a sigspec.

Basically the way the RTLIL parser works is that it reads a signed number into a C `long` in an ordinary way. After that, if this number is used in a sigspec, it is ANDed with 0xFFFFFFFF and turned into an RTLIL constant.

Besides just being inconsistent (check the examples I have in #2066), if you have an unsigned constant that's 0x80000000 or higher and you serialize it as a bare integer, then on platforms with 32-bit `long` (so basically on 32-bit platforms) this RTLIL file will fail to parse*. But it will parse just fine on 64-bit platforms. So any emitter of RTLIL files has to be careful to not produce files that will be readable on some platforms but not others--which is frankly preposterous.

\* It actually used to invoke UB and silently produce incorrect netlist until I made it not do that in commit ca70a1049ff18c8702fdc2ac4b3901685551cb28...

Reply:
Actually, it seems like I misremembered. In fact, in commit ca70a1049ff18c8702fdc2ac4b3901685551cb28 I restricted the range of bare integer constants to the domain of 32-bit signed numbers, regardless of the platform.

So I think you can just write it in the manual that bare integers in sigspecs have a range of [-2147483648, 2147483648) and anything outside of that range must be emitted as a bit vector. Also that the sign of the bare integer (when it is used in a sigspec) is completely ignored, and the resulting constant is always 32-bit.

Reply:
Thanks for the clarification. Updated with explanations in both the `<integer>` section and the `<sigspec>` section.